### DISCUSSION TITLE
counting backward with conditionals

### DISCUSSION THREAD

**Discussion by LeslieHeld** [jspsych/jsPsych#1366] [Q&A] [answered]
Hello!

I have a question concerning the ```jsPsych.data.get().last(x).values()[0].xy``` command. Namely, I'm running an experiment with a timeline including conditional statements.

In the main block the timeline is ```[loop_node,target,correct_conditional,points_conditional]```  (with ```loop_node``` looping over ```[fixation, activation conditional]``` until participants activate the fixation screen with the correct button). In the baseline block the timeline is ```[loop_node,target,correct_conditional]``` (hence the same without ```points_conditional```).

If ```correct_conditional``` or ```points_conditional``` evaluate true, participants will receive feedback on their correctness or points respectively.

At each fixation in the ```loop_node```, I want to count back to the last fixation (or target) in order to receive a choice parameter. The problem is that depending on the conditionals evaluating false or true and the phase of the experiment (main block/baseline), I have to count back 2, 3 or 4 times. 

I also tried to use ```jsPsych.data.getLastTrialData().values()[0].choice``` instead but that does not seem to work in the ```on_finish()``` part of my timeline variable.

Do you have an idea on how I can fix this?

Thanks a lot in advance!

**Comment by jodeleeuw** [comment, marked as answer]
Hi @LeslieHeld,

I recommend using the `data` parameter to tag the choice trial with some kind of unique identifier and then use the `.filter()` method when you want to grab the data.

```js
var choice = {
	data: {phase: 'choice'}
}

var loop_node = {
	loop_function: function(){
		jsPsych.data.get().filter({phase: 'choice'}).values()[0].choice
	}
}
```

I'm not quite sure which `on_finish` you are referring to in the second part of your question.

**Reply by Nega100** [reply]
Hi! @jodeleeuw  @LeslieHeld and all JsPsych Contributors!,
I think I have the same problem. I created n-back and implemented ```conditional timeline```. I intended to measure the proportion in ```corect``` answer and after a few trials decided the conditional to end experiment/other feedback. However the ```conditional timeline``` seems not run properly. I dont understand what I missed. The correct answer is below 0.6 but the message not showing up, instead showing in the end of trials. Thanks for the answer!!
```
    var cons_end_2 = {
      type: 'html-keyboard-response',
      stimulus: '<p>Too many error!</p>',
      choices: jsPsych.NO_KEYS,
      trial_duration: 5000
    };
    var conditional_2 = {
        timeline: [cons_end_2],
        conditional_function: function() {
            var two_back_cond = jsPsych.data.get().filter({even:'2-back'}).last(nback_length-2);
            var two_back_correct = two_back_cond.filter({correct: true}).count();
            var prop_tab_correct = two_back_correct/(nback_length-2)
            if (prop_tab_correct < 0.6) {
                return true;
            } else {
                return false;
            }
        }
    };
    timeline.push(conditional_2)
```

**Reply by jodeleeuw** [reply]
Hi @Nega100,

This code looks OK to me at first glance. Any error messages? Have you tried using the debugging tools to see what the value of the variables are in the `conditional_function`?

**Reply by Nega100** [reply]
Thanks Josh!. I will try look more detailed what really happening in my code. There are no error messages and in ```console.log()``` the result came out as I expected. If you have other debugging tools I'd very appreciated. Thanks a lot for the fast reply!

**Reply by jodeleeuw** [reply]
Hi @Nega100 ,

Web browsers have a great set of debugging tools built into them. Here's [an overview of how to debug in Chrome](https://developers.google.com/web/tools/chrome-devtools/javascript), and the process is very similar in other browsers.
